src/wayland/xdg_toplevel_tag.rs

```diff
@@ -0,0 +1,195 @@
+//! use wayland_protocols::xdg::shell::server::xdg_toplevel::XdgToplevel;
+//! use wayland_server::protocol::wl_surface::WlSurface;
+//! use smithay::delegate_xdg_toplevel_tag;
+//!
+//! # struct State;
+//! # let mut display = wayland_server::Display::<State>::new().unwrap();
+//!
+//! XdgToplevelTagManager::new::<State>(
+//!     &display.handle(),
+//! );
+//!
+//! impl XdgToplevelTagHandler for State {
+//!     fn on_set_tag(&mut self, toplevel: XdgToplevel, tag: String) {
+//!         dbg!(tag);
+//!     }
+//!
+//!     fn on_set_description(&mut self, toplevel: XdgToplevel, description: String) {
+//!         dbg!(description);
+//!     }
+//! }
+//!
+//! delegate_xdg_toplevel_tag!(State);
+//! ```
+
+use std::sync::{Arc, Mutex};
+
+use wayland_protocols::xdg::{
+    shell::server::xdg_toplevel::XdgToplevel,
+    toplevel_tag::v1::server::xdg_toplevel_tag_manager_v1::{self, XdgToplevelTagManagerV1},
+};
+
+use wayland_server::{
+    backend::GlobalId, Client, DataInit, Dispatch, DisplayHandle, GlobalDispatch, New, Resource,
+};
+
+use crate::wayland::{compositor, shell::xdg::XdgShellSurfaceUserData};
+
+/// Data associated with WlSurface
+/// Represents the client pending state
+///
+/// ```no_run
+/// use smithay::wayland::compositor;
+/// use smithay::wayland::xdg_toplevel_tag::XdgToplevelTagSurfaceData;
+///
+/// # let wl_surface = todo!();
+/// compositor::with_states(&wl_surface, |states| {
+///     let mut data = states.data_map.get::<XdgToplevelTagSurfaceData>().unwrap();
+///     dbg!(data.tag());
+///     dbg!(data.description());
+/// });
+/// ```
+#[derive(Default, Debug)]
+pub struct XdgToplevelTagSurfaceData {
+    tag: Mutex<Option<Arc<str>>>,
+    description: Mutex<Option<Arc<str>>>,
+}
+
+impl XdgToplevelTagSurfaceData {
+    /// Get tag of a toplevel.
+    ///
+    /// The tag may be shown to the user in UI,
+    /// so it's preferable for it to be human readable,
+    /// but it must be suitable for configuration files and should not be translated.
+    ///
+    /// Example of a tag would be "main window", "settings", "e-mail composer".
+    pub fn tag(&self) -> Option<Arc<str>> {
+        self.tag.lock().unwrap().clone()
+    }
+
+    /// Get description of a toplevel.
+    ///
+    /// This description may be shown to the user in UI or read by a screen reader for accessibility purposes.
+    pub fn description(&self) -> Option<Arc<str>> {
+        self.description.lock().unwrap().clone()
+    }
+}
+
+/// Handler trait for xdg toplevel tag events.
+pub trait XdgToplevelTagHandler:
+    GlobalDispatch<XdgToplevelTagManagerV1, ()> + Dispatch<XdgToplevelTagManagerV1, ()> + 'static
+{
+    /// Toplevel tag was set/updated.
+    #[allow(unused)]
+    fn on_set_tag(&mut self, toplevel: XdgToplevel, tag: String) {}
+
+    /// Toplevel description was set/updated.
+    #[allow(unused)]
+    fn on_set_description(&mut self, toplevel: XdgToplevel, description: String) {}
```

Comment:
It doesn't look like any trait in smithay currently has methods named with `on_*`. So it might be more consistent to just call them `set_tag` and `set_description`?

Not sure about having empty default implementations, vs just making the functions mandatory. At least for the intended purpose of the protocol, every implementer should care about the tag, though maybe not the description.

Reply:
Yeah right, we don't do `on_*`. We have `*_changed` in xdg shell tho, so perhaps that's  a more consistent choice. `set_*` is good as well as that does not indicate any deduplication logic which I don't do in this implementation.

---

The callbacks are optional, as you can still access the data without listening to those.